### PR TITLE
use default value of setEllipsize when numberOfLines higher then 1

### PR DESCRIPTION
**Pull request by fabriziobertoglio1987**
## Summary

This issue fixes https://github.com/facebook/react-native/issues/30326 fixes https://github.com/facebook/react-native/issues/30684 fixes https://github.com/facebook/react-native/issues/30272

ellipsizeMode other then tail does not work when using `numberOfLines` higher then 1 and ellipsizeMode is not clip on Android.

```jsx
<Text numberOfLines={2} ellipsizeMode="middle">Some long text</Text>
```

FURTHER INFORMATION
The Android Open Source Project `TextView` class includes explanation of the limitation of `setEllipsize`

https://github.com/aosp-mirror/platform_frameworks_base/blob/523e3b9b7922685cff53d046884e0037004dcea7/core/java/android/widget/TextView.java#L10372-L10375
https://developer.android.com/reference/android/widget/TextView#setEllipsize(android.text.TextUtils.TruncateAt)

```java
  * If {@link #setMaxLines} has been used to set two or more lines,
  * only {@link android.text.TextUtils.TruncateAt#END} and
  * {@link android.text.TextUtils.TruncateAt#MARQUEE} are supported
  * (other ellipsizing types will not do anything).
  public void setEllipsize(TextUtils.TruncateAt where) {
```

The value `tail` corresponds to `TextUtils.TruncateAt.END` in the ReactAndroid API

https://github.com/facebook/react-native/blob/eeb36f470929c2fdd8e1ed69898a5ba9144b8715/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java#L54-L56

This problem was observed by @Simek with comment https://github.com/facebook/react-native-website/pull/2518#pullrequestreview-599738809

## Changelog

[Android] [Fixed] - use default value of setEllipsize when numberOfLines > 1

## Test Plan

This changes fix the Java API for which I can not write Java tests as explained in commit https://github.com/facebook/react-native/commit/709a441ecf54cd9465f5946af0454ee7d10d5cbe
The java TextInputTest was excluded from the test suite in commit https://github.com/facebook/react-native/commit/709a441ecf54cd9465f5946af0454ee7d10d5cbe as they need the Yoga libraries to run

**<details><summary>CLICK TO OPEN TESTS RESULTS - More then 1 line</summary>**
<p>

when using some long text, numberOfLines 2 and ellipsizeMode="middle" the text is not truncated

```jsx
<Text numberOfLines={2} ellipsizeMode="middle" >Some long text</Text>
```

| **BEFORE** | 
|:-------------------------:|
|  |

the text will be truncated by default as tail in case numberOfLines > 1

```jsx
<Text numberOfLines={1} ellipsizeMode={"middle"}>
Maximum of one line no matter now much I write here. If I keep
writing it{"'"}ll just truncate after one line
</Text>
<Text ellipsizeMode={"middle"} numberOfLines={2} style={{marginTop: 20}}>
Maximum of two lines no matter now much I write here. If I keep
writing it{"'"}ll just truncate after two lines
</Text>
````

| **AFTER** | 
|:-------------------------:|
|  |

use `clip` with `numberOfLines` higher then 1 `<Text ellipsizeMode={"clip"} numberOfLines={2}}>` 

```jsx
<Text numberOfLines={1} ellipsizeMode={"clip"}>
Maximum of one line no matter now much I write here. If I keep
writing it{"'"}ll just truncate after one line
</Text>
<Text ellipsizeMode={"clip"} numberOfLines={2} style={{marginTop: 20}}>
Maximum of two lines no matter now much I write here. If I keep
writing it{"'"}ll just truncate after two lines
</Text>
```

| **AFTER** | 
|:-------------------------:|
|  |

</p>
</details>

**<details><summary>CLICK TO OPEN TESTS RESULTS - Only 1 line</summary>**
<p>

Testing ellipsizeMode={"clip"}, the `...` are not displayed

```jsx
<Text numberOfLines={1} ellipsizeMode={"clip"}>
Maximum of one line no matter now much I write here. If I keep
writing it{"'"}ll just truncate after one line
</Text>
```

| **AFTER** | 
|:-------------------------:|
|  |

Testing ellipsizeMode={"middle"}, the `...` are displayed in the middle

```jsx
<Text numberOfLines={1} ellipsizeMode={"middle"}>
Maximum of one line no matter now much I write here. If I keep
writing it{"'"}ll just truncate after one line
</Text>
```

| **AFTER** | 
|:-------------------------:|
|  |

Testing ellipsizeMode={"tail"}, the `...` are displayed at the end

```jsx
<Text numberOfLines={1} ellipsizeMode={"tail"}>
Maximum of one line no matter now much I write here. If I keep
writing it{"'"}ll just truncate after one line
</Text>
```

| **AFTER** | 
|:-------------------------:|
|  |

</p>
</details>

**<details><summary>CLICK TO OPEN TESTS RESULTS - Updating numberOfLines with `ellipsizeMode` in `middle`</summary>**
<p>

From code review https://github.com/facebook/react-native/pull/31064#discussion_r639942347

TEST SCENARIO
- The user updates the prop `numberOfLines` from `1` to `2`. The `Text` was previously set with `ellipsizeMode` `head` (`TextUtils.TruncateAt.START`) or `middle` (`TextUtils.TruncateAt.MIDDLE`). The value of prop `ellipsizeMode` is not changed.

RESULT
- The `Text` style `ellipsizeMode` is automatically changed from `head` or `middle` to `tail` (`TextUtils.TruncateAt.END`) in ReactAndroid as by requirement.

```javascript
class TextWithNumberOfLines extends React.Component<{...}, $FlowFixMeState> {
  state = {numberOfLines: 1};
  increaseLines = () => {
    this.setState(state => ({
      numberOfLines: state.numberOfLines + 1,
    }));
  };

  render() {
    const {numberOfLines} = this.state;
    return (
      <View>
        <Text
          numberOfLines={numberOfLines}
          ellipsizeMode="middle"
          style={{marginTop: 20}}>
          ellipsizeMode is middle and numberOfLines is {numberOfLines} Maximum
          of two lines no matter now much I write here. If I keep writing it
          {"'"}ll just truncate after two lines
        </Text>
        <View
          style={{
            flexDirection: 'row',
            justifyContent: 'space-between',
            marginTop: 15,
          }}>
          <Button title="increase numberOfLines" onPress={this.increaseLines} />
          <Button
            title="reset"
            onPress={() => this.setState({numberOfLines: 1})}
          />
        </View>
      </View>
    );
  }
}
```

<video src="https://user-images.githubusercontent.com/24992535/119756805-965eb500-bed6-11eb-9489-bb5050a52293.mp4" width="700" height="" />

</p>
</details>

**<details><summary>CLICK TO OPEN TESTS RESULTS - Updating numberOfLines with `ellipsizeMode` in `clip`</summary>**
<p>

TEST SCENARIO
- The above test case scenario with `ellipsizeMode` in `clip`

RESULT
- The `Text` style `ellipsizeMode` is kept as `clip` when multiline

```javascript
class TextWithNumberOfLines extends React.Component<{...}, $FlowFixMeState> {
  state = {numberOfLines: 1};
  increaseLines = () => {
    this.setState(state => ({
      numberOfLines: state.numberOfLines + 1,
    }));
  };

  render() {
    const {numberOfLines} = this.state;
    return (
      <View>
        <Text
          numberOfLines={numberOfLines}
          ellipsizeMode="clip"
          style={{marginTop: 20}}>
          ellipsizeMode is clip and numberOfLines is {numberOfLines} Maximum of
          two lines no matter now much I write here. If I keep writing it
          {"'"}ll just truncate after two lines
        </Text>
        <View
          style={{
            flexDirection: 'row',
            justifyContent: 'space-between',
            marginTop: 15,
          }}>
          <Button title="increase numberOfLines" onPress={this.increaseLines} />
          <Button
            title="reset"
            onPress={() => this.setState({numberOfLines: 1})}
          />
        </View>
      </View>
    );
  }
}
```

<video src="https://user-images.githubusercontent.com/24992535/119757792-10436e00-bed8-11eb-9fe7-cd1e05b9704f.mp4" width="700" height="" />

</p>
</details>

### CODE REVIEW

**Comment by fabriziobertoglio1987** [comment]
- [ ] fix not possible to set ellipsizeMode clip on Text with numberOfLines 2



:construction: does not work on all scenarios :construction:

**<details><summary>CLICK TO OPEN TESTS RESULTS</summary>**
<p>

not possible to set ellipsizeMode clip on Text with numberOfLines 2

```jsx
<Text numberOfLines={1} ellipsizeMode={"clip"}>
Maximum of one line no matter now much I write here. If I keep
writing it{"'"}ll just truncate after one line
</Text>
<Text ellipsizeMode={"clip"} numberOfLines={2} style={{marginTop: 20}}>
Maximum of two lines no matter now much I write here. If I keep
writing it{"'"}ll just truncate after two lines
</Text>
```

| **AFTER** | 
|:-------------------------:|
|  |

</p>
</details>

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java [hunk 1]
```diff
@@ -42,16 +42,19 @@
     Spacing.ALL, Spacing.LEFT, Spacing.RIGHT, Spacing.TOP, Spacing.BOTTOM,
   };
   private static final String TAG = "ReactTextAnchorViewManager";
+  private Integer mNumberOfLines = ViewDefaults.NUMBER_OF_LINES;
```
Comment (mdvacca) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java:45]: NIT: can you use int instead of Integer here?

**Review by mdvacca — APPROVED** [review]
Thanks for working on this @fabriziobertoglio1987 

Code looks good to me, I just added a comment. 

If you can update the code I will import the PR

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java (hunk 1 above)
Comment (fabriziobertoglio1987) [reply ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java]: Thanks a lot @mdvacca I added the change with commit https://github.com/facebook/react-native/pull/31064/commits/f602cd9e90cc45b52fa64a43013c4a5c44370f65 I wish you a good day

**Review by mdvacca — APPROVED** [review]
lgtm

**Comment by mdvacca** [comment]
Importing the PR

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java [hunk 2]
```diff
@@ -42,16 +42,19 @@
...
   @ReactProp(name = ViewProps.NUMBER_OF_LINES, defaultInt = ViewDefaults.NUMBER_OF_LINES)
   public void setNumberOfLines(ReactTextView view, int numberOfLines) {
+    mNumberOfLines = numberOfLines;
     view.setNumberOfLines(numberOfLines);
   }
 
   @ReactProp(name = ViewProps.ELLIPSIZE_MODE)
   public void setEllipsizeMode(ReactTextView view, @Nullable String ellipsizeMode) {
-    if (ellipsizeMode == null || ellipsizeMode.equals("tail")) {
+    boolean useDefaultValue = mNumberOfLines > 1 && !ellipsizeMode.equals("clip");
```
Comment (mdvacca) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java:56]: Actually, @fabriziobertoglio1987 can we retrieve NumberOfLines from the view?

View Managers shouldn't keep state

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java [hunk 3]
```diff
@@ -42,16 +42,19 @@
...
+    if (ellipsizeMode == null || useDefaultValue || ellipsizeMode.equals("tail")) {
```
[the remark this answers is not in the document]
Comment (fabriziobertoglio1987) [reply ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java]: Thanks a lot David and Andrei

I pushed https://github.com/facebook/react-native/pull/31064/commits/a271d11c61b6e3af3c78ccdc53de44c5bd93b114 and I run the following test cases which are still **IN PROGRESS**, I will further review this changes and update this discussion in 5-6 hours. :pray: 

**<details><summary>CLICK TO OPEN TESTS RESULTS - Updating numberOfLines with `ellipsizeMode` in `middle`</summary>**
<p>

From code review https://github.com/facebook/react-native/pull/31064#discussion_r639942347

TEST SCENARIO
- The user updates the prop `numberOfLines` from `1` to `2`. The `Text` was previously set with `ellipsizeMode` `head` (`TextUtils.TruncateAt.START`) or `middle` (`TextUtils.TruncateAt.MIDDLE`). The value of prop `ellipsizeMode` is not changed.

RESULT
- The `Text` style `ellipsizeMode` is automatically changed from `head` or `middle` to `tail` (`TextUtils.TruncateAt.END`) in ReactAndroid as by requirement.

```javascript
class TextWithNumberOfLines extends React.Component<{...}, $FlowFixMeState> {
  state = {numberOfLines: 1};
  increaseLines = () => {
    this.setState(state => ({
      numberOfLines: state.numberOfLines + 1,
    }));
  };

  render() {
    const {numberOfLines} = this.state;
    return (
      <View>
        <Text
          numberOfLines={numberOfLines}
          ellipsizeMode="middle"
          style={{marginTop: 20}}>
          ellipsizeMode is middle and numberOfLines is {numberOfLines} Maximum
          of two lines no matter now much I write here. If I keep writing it
          {"'"}ll just truncate after two lines
        </Text>
        <View
          style={{
            flexDirection: 'row',
            justifyContent: 'space-between',
            marginTop: 15,
          }}>
          <Button title="increase numberOfLines" onPress={this.increaseLines} />
          <Button
            title="reset"
            onPress={() => this.setState({numberOfLines: 1})}
          />
        </View>
      </View>
    );
  }
}
```

<video src="https://user-images.githubusercontent.com/24992535/119756805-965eb500-bed6-11eb-9489-bb5050a52293.mp4" width="700" height="" />

</p>
</details>

**<details><summary>CLICK TO OPEN TESTS RESULTS - Updating numberOfLines with `ellipsizeMode` in `clip`</summary>**
<p>

TEST SCENARIO
- The above test case scenario with `ellipsizeMode` in `clip`

RESULT
- The `Text` style `ellipsizeMode` is kept as `clip` when multiline

```javascript
class TextWithNumberOfLines extends React.Component<{...}, $FlowFixMeState> {
  state = {numberOfLines: 1};
  increaseLines = () => {
    this.setState(state => ({
      numberOfLines: state.numberOfLines + 1,
    }));
  };

  render() {
    const {numberOfLines} = this.state;
    return (
      <View>
        <Text
          numberOfLines={numberOfLines}
          ellipsizeMode="clip"
          style={{marginTop: 20}}>
          ellipsizeMode is clip and numberOfLines is {numberOfLines} Maximum of
          two lines no matter now much I write here. If I keep writing it
          {"'"}ll just truncate after two lines
        </Text>
        <View
          style={{
            flexDirection: 'row',
            justifyContent: 'space-between',
            marginTop: 15,
          }}>
          <Button title="increase numberOfLines" onPress={this.increaseLines} />
          <Button
            title="reset"
            onPress={() => this.setState({numberOfLines: 1})}
          />
        </View>
      </View>
    );
  }
}
```

<video src="https://user-images.githubusercontent.com/24992535/119757792-10436e00-bed8-11eb-9fe7-cd1e05b9704f.mp4" width="700" height="" />

</p>
</details>

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java [hunk 4]
```diff
@@ -47,11 +47,16 @@
   @ReactProp(name = ViewProps.NUMBER_OF_LINES, defaultInt = ViewDefaults.NUMBER_OF_LINES)
   public void setNumberOfLines(ReactTextView view, int numberOfLines) {
     view.setNumberOfLines(numberOfLines);
+    boolean isMultiline = view.getNumberOfLines() > 1 && view.getNumberOfLines() != ViewDefaults.NUMBER_OF_LINES;
```
[the remark this answers is not in the document]
Comment (fabriziobertoglio1987) [reply ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java]: https://github.com/facebook/react-native/pull/31064/commits/c650f35417d95fbca0fb3185d45fff6b23291906

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java [hunk 5]
```diff
@@ -47,11 +47,16 @@
...
+    if (isMultiline && view.getEllipsizeLocation() != null) {
+      view.setEllipsizeLocation(TextUtils.TruncateAt.END);
+    }
   }
 
   @ReactProp(name = ViewProps.ELLIPSIZE_MODE)
   public void setEllipsizeMode(ReactTextView view, @Nullable String ellipsizeMode) {
-    if (ellipsizeMode == null || ellipsizeMode.equals("tail")) {
+    boolean isMultiline = view.getNumberOfLines() > 1 && view.getNumberOfLines() != ViewDefaults.NUMBER_OF_LINES;
```
[the remark this answers is not in the document]
Comment (fabriziobertoglio1987) [reply ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java]: https://github.com/facebook/react-native/pull/31064/commits/c650f35417d95fbca0fb3185d45fff6b23291906

**Comment by gabrieldonadel** [comment]
Any updates on this @mdvacca @fabriziobertoglio1987 ?

**Pull request closed without merging**